include/souper/Infer/AbstractInterpreter.h

```diff
@@ -16,29 +16,75 @@
-    llvm::KnownBits or_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits xor_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits shl(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits lshr(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ashr(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits eq(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ne(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ult(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits slt(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ule(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits sle(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
+    llvm::KnownBits add(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits addnsw(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits sub(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits subnsw(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits mul(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits udiv(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits urem(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits and_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits or_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits xor_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits shl(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits lshr(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ashr(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits eq(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ne(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ult(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits slt(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ule(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits sle(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
   }
+
+  bool isConcrete(souper::Inst *I,
+                  bool ConsiderConsts = true,
+                  bool ConsiderHoles = true);
+
+  class KnownBitsAnalysis {
+    std::unordered_map<Inst*, llvm::KnownBits> KBCache;
+
+    // Checks the cache or instruction metadata for knonwbits information
+    bool cacheHasValue(Inst *I);
+
+  public:
+    llvm::KnownBits findKnownBits(Inst *I,
+                                  ConcreteInterpreter &CI,
+                                  bool PartialEval = true);
```

Comment:
is this default value really needed? I really like to avoid default valued boolean flags when possible, they make using functions confusing

Reply:
The only way I have used this till now is to test this thing when running experiments last month. This was already there before my modifications though. @manasij7479 do you've any comments?